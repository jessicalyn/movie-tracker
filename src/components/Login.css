.error-message{
    color:white;
}

form input{
    height: 40px;
    font-size: 20px;
    width: 100%;
}

form button{
    height: 45px;
    margin-bottom: 20px;
    font-size: 30px; 
    text-transform: uppercase;
    border: 3px solid #00d7c3;
    border-radius: 50px;
    background-color: white;
}

form button:hover{
    background-color: #FFCC33;
}

form button:focus {
    outline:0;
}

.input-box{
    display: flex;
    padding: 10px;
    background-color: white;
    margin-bottom: 22px;
    justify-content: space-between;
}

.input-box input:focus {
    outline: 0;
}

.input-box img {
    width: 45px;
    height: 48px;
}

.login {
    display: flex;
    background-color: #333;
    background: url("../images/login-background.jpeg");
    background-size: cover;
    height: 700px;
    width: 100%;
    display: inline-block;
    justify-content: center;
    align-items: center;
}

.login h3{
    color: white;
    border-bottom:6ps solid white;
}

.login-box {
    width: 30%;
    background-color: black;
    opacity:0.5;
    margin: 0 auto;
    position: relative;
    top: 50%;
    left: 18%;
    height: 500px;
    padding: 20px;
    transform: translate(-50%,-50%);
}

.login-box form {
    display: flex;
    flex-direction: column;
}

.login-header h3 {
    font-size: 23px;
}

.password-link{
    color:white;
    text-decoration: underline;    
}

