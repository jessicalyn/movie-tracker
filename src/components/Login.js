import { Route, Redirect } from 'react-router'
import React from "react";
import { Component } from "react"
import { loginUser } from '../Actions/index'
import { connect } from 'react-redux'


export class Login extends Component {
  constructor() {
    super();
    this.state = {
      email: "",
      password: "",
      error: ""
    };
  }

  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({ [name]: value })
  }

  validateUser = async (e) => {
    e.preventDefault()
    const { email, password} = this.state
    const url = 'http://localhost:3000/api/users'
    const data = { email: email.toLowerCase(), password: password }
    try {
      const checkUser = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(data),
        headers:{
          'Content-Type': 'application/json'
        }
      })
      const response = await checkUser.json()
      if(response.status === "success"){
        return this.props.loginUser(response.data.id)
      }
    } catch(error) {
      return this.setState({ error: "Email and Password do not match. Please try again or Signup."})
    }
  }

  render() {
    return (
      <div className="login">
        <form onSubmit={this.validateUser}>
          <input
            type="text"
            name="email"
            value={this.state.email}
            id="email-input"
            onChange={this.handleChange}
          />
          <input
            type="text"
            name="password"
            id="password-input"
            value={this.state.password}
            onChange={this.handleChange}
          />
          <button>Submit</button>
          {this.state.error}
        </form>
        <Route exact path='/login' render={() => (
          this.props.user.id && <Redirect to="/"/>
        )} />
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  loginUser: (email) => dispatch(loginUser(email))
})

export default connect(mapStateToProps, mapDispatchToProps)(Login)