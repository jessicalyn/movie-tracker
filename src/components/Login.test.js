import React from 'react';
import ReactDOM from 'react-dom';
import { Login, mapStateToProps, mapDispatchToProps, } from './Login'
import { shallow } from 'enzyme';

describe ('Login', () => {
    describe('Login', () => {
        let wrapper;

        beforeEach(() => {
            wrapper = shallow(<Login/>)
        })

        it('should match the snapshot with all data passed in', () => {
            expect(wrapper).toMatchSnapshot()
        })
        it('should update email when handle change is invoked', () => {
           const mockEvent=  {
                target: {
                    name: 'email',
                    value:'abc@gmail.com'
                }
            }
            expect(wrapper.state("email")).toEqual("");
            wrapper.instance().handleChange(mockEvent);
            expect(wrapper.state("email")).toEqual("abc@gmail.com");
        })
        it("should update password when handle change is invoked", ()=> {
            const mockEvent = {
                target: {
                    name:'password',
                    value: '1234'
                }
            }
            expect(wrapper.state("password")).toEqual("")
            wrapper.instance().handleChange(mockEvent);
            expect(wrapper.state("password")).toEqual("1234")
            
        })

        it("it should validate user on submit", async () => {

            const mockData = {
                email: "abc@gmail.com",
                password: "1234"
            }
            wrapper.instance().setState({email: mockData.email, password: mockData.password})
            const expectedUrl = 'http://localhost:3000/api/users'
            const mockOptions = {
                method: 'POST',
                body: JSON.stringify(mockData),
                headers:{
                  'Content-Type': 'application/json'
                }
              }
            const mockEvent =  {preventDefault: () => {}}

            window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
                ok: true,
                status: 200,
                json: () => Promise.resolve(mockData),
              }));
            wrapper.instance().validateUser(mockEvent)  
            expect(window.fetch).toHaveBeenCalledWith(expectedUrl, mockOptions)
        })

        it("should login user if response status is success", () => {
            const mockData = {
                email: "niroz.aryal11@gmail.com",
                password: "password"
            }
            wrapper.instance().setState({email: mockData.email, password: mockData.password})
            const mockLoginUser = jest.fn()
            const expectedUrl = 'http://localhost:3000/api/users'
            const mockOptions = {
                method: 'POST',
                body: JSON.stringify(mockData),
                headers:{
                  'Content-Type': 'application/json'
                }
              }
            const mockEvent =  {preventDefault: () => {}}

            window.fetch = jest.fn()
            .mockImplementation(() => Promise.resolve({
                ok: false,
                status: 400,
                json: () => Promise.resolve(mockData),
              }));
            wrapper.instance().validateUser(mockEvent)  
            expect(window.fetch).toHaveBeenCalledWith(expectedUrl, mockOptions)
            wrapper.instance().setState({error: "Email and Password do not match. Please try again or Signup."})
        })
            
            



            
    })
})
