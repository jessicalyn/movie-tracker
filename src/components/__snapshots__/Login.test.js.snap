// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login should match the snapshot with all data passed in 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
        >
          <input
            id="email-input"
            name="email"
            onChange={[Function]}
            type="text"
            value=""
          />
          <input
            id="password-input"
            name="password"
            onChange={[Function]}
            type="text"
            value=""
          />
          <button>
            Submit
          </button>
          
        </form>,
        <Route
          exact={true}
          path="/login"
          render={[Function]}
        />,
      ],
      "className": "login",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              id="email-input"
              name="email"
              onChange={[Function]}
              type="text"
              value=""
            />,
            <input
              id="password-input"
              name="password"
              onChange={[Function]}
              type="text"
              value=""
            />,
            <button>
              Submit
            </button>,
            "",
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "email-input",
              "name": "email",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "password-input",
              "name": "password",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          "",
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/login",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
          >
            <input
              id="email-input"
              name="email"
              onChange={[Function]}
              type="text"
              value=""
            />
            <input
              id="password-input"
              name="password"
              onChange={[Function]}
              type="text"
              value=""
            />
            <button>
              Submit
            </button>
            
          </form>,
          <Route
            exact={true}
            path="/login"
            render={[Function]}
          />,
        ],
        "className": "login",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                id="email-input"
                name="email"
                onChange={[Function]}
                type="text"
                value=""
              />,
              <input
                id="password-input"
                name="password"
                onChange={[Function]}
                type="text"
                value=""
              />,
              <button>
                Submit
              </button>,
              "",
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "email-input",
                "name": "email",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "password-input",
                "name": "password",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
            "",
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/login",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
